doc/smart-switch/pmon/smartswitch-pmon.md

```diff
@@ -0,0 +1,1025 @@
+* The get_reboot_cause will return the current reboot-cause of the module.
+* For persistent storage of the DPU reboot-cause and reboot-caue-history files use the existing host storage path and mechanism.
+
+#### Schema for REBOOT_CAUSE - switch stateDB
+```
+  Key: "REBOOT_CAUSE|2023_06_18_14_56_12"
+
+  "REBOOT_CAUSE|2023_06_18_14_56_12": {
+    "value": {
+      "cause": "REBOOT_CAUSE_HOST_RESET_DPU",
+      "comment": "N/A",
+      "device": "DPU5",
+      "time": "2023_06_18_14_56_12",
+      "user": "N/A"
+    }
+  },
+
+```
+2. Though the get_oper_status(self) can get the operational status of the DPU Modules, the current implementation only has limited capabilities.
+    * Can only state MODULE_STATUS_FAULT and can't show exactly where in the state progression the DPU failed. This is critical in fault isolation, DPU switchover decision, resiliency and recovery
+    * Though this is platform implementation specific, in a multi vendor use case, there has to be a consistent way of storing and accessing the information.
+    * Store the state progression (dpu_midplane_link_state, dpu_control_plane_state, dpu_data_plane_state) on the host ChassisStateDB.
+    * get_state_info(self) will return an object with the ChassisStateDB data
+    * Potential consumers: HA, LB, Switch CLIs, Utils (install/repair images), Life Cycle Manager 
+    * Use cases: HA, Debuggability, error recovery (reset, power cycle) and fault management, consolidated view of Switch and DPU state
+#### DPU_STATE definition
+dpu_midplane_link_state: up refers to the pcie link between the NPU and DPU is operational. This will be updated by the switch pcied.
+
+dpu_control_plane_state: up  refers to DPU SONIC booted up, all the containers are up, the interfaces are up and dpu is ready to take SDN policy update
+
+dpu_data_plane_state: up  refers to configuration downloaded, the pipeline stages are up and DPU hardware (port/ASIC) is ready to take traffic
+
+```
+    ChassisStateDB Schema for DPU_STATE
+
+    Table: “DPU_STATE”
+
+    SCHEMA
+    key:  DPU_STATE|DPU0
+
+        "id": "1",
+        ”dpu_midplane_link_state”: "up"
+        “dpu_midplane_link_time": "timestamp",
+        "dpu_midplane_link_reason": "up_down_related string",
+        ”dpu_control_plane_state": "up"
+        ”dpu_control_plane_time": ”timestamp",
+        ”dpu_control_plane_reason": ”containers restarting",
+        ”dpu_data_plane_state": ”down",
+        ”dpu_data_plane_time": ”timestamp",
+        ”dpu_data_plane_reason": ”Pipeline failure",
+```
+
+3. Each DPU has to store the health data in its local DB and should provide it to the switch.
+* When the "show system-health ..." CLI is executed on the switch, the "UserDefinedChecker" class will collect the data and feed it to the CLI. It is up to the platform on how this is done.  However, for faster access store it in the switch ChassisStateDB.
```

Comment:
as chatted offline, let's add a notification channel between some component in PMON to hamgrd for notifying the state being changed, as these are critical health signal HA is relying on.

Reply:
@r12f hamgrd can subscribe to the redis db table events to get these notifications

Reply:
Sure, I'll add another document to detail out the interactions between the hamgrd and PMON dpu_state change handler.